### ISSUE TITLE
sql: `regexp_replace` does not properly support newline-sensitivity

### ISSUE THREAD

**Issue by sploiselle** [cockroachdb/cockroach#12362]
Full newline-sensitivity through `regexp_replace` (in terms of Postgres' [support for the feature](https://www.postgresql.org/docs/9.1/static/functions-matching.html#POSIX-MATCHING-RULES)), enables you to use `^` and `$` to match characters before/after newlines with the `n` flag. However, Cockroach is not currently finding those matches.

### Steps to Repro
```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 'n');
```

### Expected Behavior
Return
```
+--------------------------------+
| regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 'n') |
+--------------------------------+
| DOGGIE␤                        |
| CAT ␤                          |
| DOG                            |
+--------------------------------+
```
### Actual Behavior
```
+--------------------------------+
| regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 'n') |
+--------------------------------+
| DOGGIE␤                        |
| dog ␤                          |
| DOG                            |
+--------------------------------+
```

#### More info

FWIW, in Postgres:
```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 'n');
```
```
 regexp_replace
----------------
 DOGGIE        +
 CAT           +
 DOG
```

**Comment by sploiselle** [comment]
tl;dr: CockroachDB's default behavior is "Weird" inverse partial newline-sensitivity, and each of the newline-sensitivity flags exhibit unexpected behavior.

## Newline-Insensitive (`s`)

From [Postgres' docs](https://www.postgresql.org/docs/9.1/static/functions-matching.html#POSIX-MATCHING-RULES) on newline-sensitivity:

- `.` matches newline characters
- `^` matches only beginning of string

### Postgres Example

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 's') AS PSQL_Insensitive;
```
```
psql_insensitive
------------------
DOGGIE          +
dog             +
DOG
```

### CRDB Example

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 's') AS CRDB_Insensitive;
```
```
+------------------+
| CRDB_Insensitive |
+------------------+
| DOGGIE␤          |
| CAT              |
+------------------+
```

Despite being newline-insensitive, CRDB matches `^d` as the beginning of a line, though that feature should be disabled.

## Newline-Sensitivity (`n`)

From Postgres' docs:

- `.` *does not* match newline characters
- `^` matches the empty space after a newline character

### Postgres

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 'n') AS PSQL_Sensitive;
```
```
psql_sensitive
----------------
DOGGIE        +
CAT           +
DOG
```

### CRDB

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 'n') AS CRDB_Sensitive;
```
```
+-------------------+
|  CRDB_Sensitive   |
+-------------------+
| DOGGIE␤           |
| dog ␤             |
| DOG               |
+-------------------+
```

Despite being newline-sensitive, CRDB does not match `^d` after the newline character. However, CRDB succeeds in that `.` does not match newline characters :

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', 'd.+', 'CAT', 'n') AS CRDB_Sensitive_2;
```
```
+------------------+
|  CRDB_Sensitive  |
+------------------+
| DOGGIE␤          |
| CAT␤             |
| DOG              |
+------------------+
```


## Partial Newline-Sensitivity (`p`)

From Postgres' docs:

- `.` *does not* match newline characters
- `^` matches only beginning of string

### Postgres

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^D.+', 'CAT', 'p') AS PSQL_Partial;
```
```
psql_partial
--------------
CAT         +
dog         +
DOG
```

### CRDB

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^D.+', 'CAT', 'p') AS CRDB_Partial;
```
```
+--------------+
| CRDB_Partial |
+--------------+
| CAT          |
+--------------+
```

CockroachDB fails to honor the `.` not matching newline character rules. However, it does succeed in disabling `^` matching the empty space after newlines:

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 'p') AS CRDB_Partial_2;
```
```
+-------------------+
|  CRDB_Partial_2   |
+-------------------+
| DOGGIE␤           |
| dog ␤             |
| DOG               |
+-------------------+
```

## Inverse Partial/Weird (`w`)

From Postgres' docs:

- `.` matches newline characters
- `^` matches the empty space after a newline character

### Postgres

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 'w') AS Postgres_Weird;
```
```
postgres_weird
----------------
DOGGIE        +
CAT
```

### CRDB

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT', 'w') AS CRDB_Weird;
```
```
+-------------+
| CRDB_Weird  |
+-------------+
| DOGGIE␤     |
| CAT         |
+-------------+
```

## Default

### Postgres

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT') AS Postgres_Default;
```
```
postgres_default
------------------
DOGGIE          +
dog             +
DOG
```

Postgres' default behavior is the same as Newline-Insensitive.

### CRDB

```
SELECT regexp_replace(e'DOGGIE\ndog \nDOG', '^d.+', 'CAT') AS CRDB_Default;
```
```
+--------------+
| CRDB_Default |
+--------------+
| DOGGIE␤      |
| CAT          |
+--------------+
```

Cockroach's default behavior is the same as its and Postgres' "Weird" matching.
